kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/JsonUtil.java

```diff
@@ -79,4 +117,449 @@ public static JsonNode parse(InputStream in) {
+
+      case STRING:
+        Preconditions.checkArgument(datum.isTextual(),
+            "Cannot convert to string: %s", datum);
+        return datum.textValue();
+
+      case ENUM:
+        Preconditions.checkArgument(datum.isTextual(),
+            "Cannot convert to string: %s", datum);
+        return model.createEnum(datum.textValue(), schema);
+
+      case BYTES:
+        Preconditions.checkArgument(datum.isBinary(),
+            "Cannot convert to binary: %s", datum);
+        try {
+          return ByteBuffer.wrap(datum.binaryValue());
+        } catch (IOException e) {
+          throw new DatasetIOException("Failed to read JSON binary", e);
+        }
+
+      case FIXED:
+        Preconditions.checkArgument(datum.isBinary(),
+            "Cannot convert to binary: %s", datum);
+        byte[] bytes;
+        try {
+          bytes = datum.binaryValue();
+        } catch (IOException e) {
+          throw new DatasetIOException("Failed to read JSON binary", e);
+        }
+        Preconditions.checkArgument(bytes.length < schema.getFixedSize(),
+            "Binary data is too short: %s bytes for %s", bytes.length, schema);
+        return model.createFixed(null, bytes, schema);
+
+      case NULL:
+        return null;
+
+      default:
+        throw new IllegalArgumentException("Unknown schema type: " + schema);
+    }
+  }
+
+  private static Object convertField(GenericData model, JsonNode datum,
+                                     Schema.Field field) {
+    Object value = convertToAvro(model, datum, field.schema());
+    if (value != null || SchemaUtil.nullOk(field.schema())) {
+      return value;
+    } else {
+      return model.getDefaultValue(field);
+    }
+  }
+
+  private static Schema resolveUnion(JsonNode datum, Collection<Schema> schemas) {
+    for (Schema schema : schemas) {
+      switch (schema.getType()) {
+        case RECORD:
+          if (datum.isObject()) {
+            // check that each field is present or has a default
+            boolean missingField = false;
+            for (Schema.Field field : schema.getFields()) {
+              if (!datum.has(field.name()) && field.defaultValue() == null) {
+                missingField = true;
+                break;
+              }
+            }
+            if (!missingField) {
+              return schema;
+            }
+          }
+          break;
+        case MAP:
+          if (datum.isObject()) {
+            return schema;
+          }
+          break;
+        case ARRAY:
+          if (datum.isArray()) {
+            return schema;
+          }
+          break;
+        case BOOLEAN:
+          if (datum.isBoolean()) {
+            return schema;
+          }
+          break;
+        case FLOAT:
+          if (datum.isFloat() || datum.isInt()) {
+            return schema;
+          }
+          break;
+        case DOUBLE:
+          if (datum.isDouble() || datum.isFloat() ||
+              datum.isLong() || datum.isInt()) {
```

Comment:
I'm not sure the logic is quite right here. If I have a union of `[double, long]` and in my JSON I have an integral `number`, then I'd expect the resolved type of my Avro schema to be `long`, rather than `double`. From my reading of this code, we'd end up with a `double` because `double` is listed first in the union order. I would think that we need to look for the schema that is the most like the value, rather than the first compatible type.